com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DeleteApplicationRoutesStep.java

```diff
@@ -25,19 +27,24 @@
 
 @Named("deleteApplicationRoutesStep")
 @Scope(BeanDefinition.SCOPE_PROTOTYPE)
-public class DeleteApplicationRoutesStep extends UndeployAppStep {
+public class DeleteApplicationRoutesStep extends UndeployAppStep implements BeforeHooksPhaseProvider {
 
+    @Inject
     private ApplicationRoutesGetter applicationRoutesGetter;
-
     @Inject
-    public DeleteApplicationRoutesStep(ApplicationRoutesGetter applicationRoutesGetter) {
-        this.applicationRoutesGetter = applicationRoutesGetter;
-    }
+    private MtaMetadataParser mtaMetadataParser;
+    @Inject
+    private EnvMtaMetadataParser envMtaMetadataParser;
 
     @Override
-    protected StepPhase undeployApplication(CloudControllerClient client, CloudApplication cloudApplicationToUndeploy) {
+    protected StepPhase undeployApplication(CloudControllerClient client, CloudApplication cloudApplicationToUndeploy,
+                                            ProcessContext context) {
+        HooksProcessor hooksProcessor = getHooksProcessor(context);
```

Comment:
This refactoring introduces a lot of copy-pasted code. Meaning that if I want to enable the hooks in another step, I need to copy-paste this code which does all the magic. My point is that we need to make external logic which should handles this before and after step hooks calculation. Also this logic should use the newly introduced interfaces in order to operate. Meaning that when some step is started, if it is "specified" that it supports hooks, the hook building logic should understand that this step is either of type "AfterStepHookPhaseProvider" and/or "BeforeStepHookPhaseProvider" and calculate the hooks based on the current step phase.
We could discuss it - the design and the whole concept.
Try thinking of a pattern which is executing some logic by calling some delegates in order to do its job.